srml/babe/src/lib.rs

```diff
@@ -251,6 +291,34 @@ impl<T: Trait> Module<T> {
 		UnderConstruction::mutate(|z| z.iter_mut().zip(vrf_output).for_each(|(x, y)| *x^=y))
 	}
 
+	fn do_initialize() {
+		for digest in Self::get_inherent_digests()
+			.logs
+			.iter()
+			.filter_map(|s| s.as_pre_runtime())
+			.filter_map(|(id, mut data)| if id == BABE_ENGINE_ID {
+				RawBabePreDigest::decode(&mut data)
+			} else {
+				None
+			})
+		{
+			// since do_initialize can be called twice (if session module is present)
+			// => let's ensure that we only modify the storage once per block by checking
+			// current slot number
+			let current_slot = CurrentSlot::get();
+			if current_slot != digest.slot_number {
```

Comment:
Could you, please, verify that this is correct? If there could be two blocks with the same `digest.slot_number`, this condition should be fixed.